Implement federation API server authentication e2e tests.

madhusudancs opened the pull request:
This PR depends on #30397. Please review only the last commit here.

Fixes: Issue #28602.

cc @kubernetes/sig-cluster-federation

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30409)

<!-- Reviewable:end -->

madhusudancs committed:
Implement federation API server authentication e2e tests.

quinton-hoole commented:
Beautiful!  Thanks @madhusudancs .  LGTM.

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented:
@k8s-bot unit test this issue: #IGNORE

nikhiljindal commented on test/e2e/federation-authn.go:
```diff
@@ -0,0 +1,110 @@
+
+			contexts := f.GetUnderlyingFederatedContexts()
+
+			fcs, err := invalidAuthFederationClientSet(contexts[0].User)
```
Are you assuming here that contexts[0] is not for federation cluster. Is creating and passing a fake User more robust?

madhusudancs replied:
context[0] cannot be a federation API server. If it is, our tests don't work. `contexts` is obtained from calling `f.GetUnderlyingFederatedContexts()` which in turn checks that the contexts it returns are not federation API server's context/.

nikhiljindal replied:
oh, yes. so we can use any context. makes sense.
Would be good if you can add that as a comment.

nikhiljindal commented on test/e2e/federation-authn.go:
```diff
@@ -0,0 +1,110 @@
+		config.Username = ""
+	}
+
+	c, err := federation_release_1_4.NewForConfig(config)
```
We do the same in test/e2e/framework/util.go.LoadFederationClientset_1_4() after generating the config. Will be better to refactor it and reuse code.

madhusudancs replied:
There are two lines (excluding the repetitive error checks) that are common to this function and `test/e2e/framework/util.go.LoadFederationClientset_1_4()`. I don't find it worth refactoring it, esp. proliferating the number of one-time use exported functions in util.go. I would just keep them separate.

madhusudancs replied:
@nikhiljindal there is another reason I did not refactor `LoadFederationClientset_1_4()` which I forgot to mention. I need to explicitly set `config.{Password, BearerToken, Username}` to empty strings for one of the tests. It is too cumbersome to do it after the `ClientSet` is created because every client in the set has a copy of the config. If I have to set these values after the `ClientSet` is created, I have to update all of them, which is a lot more code and less neater than this.

If you insist I will be happy to export `setTimeouts` function and re-use it.

nikhiljindal replied:
Yes. Thats why I suggested refactoring the part after you have the config. Something like LoadFederationClientset14FromConfig(config). So LoadFederationClientset_1_4 becomes:
```
LoadFederationClientset_1_4() {
    config, err := LoadFederatedConfig(&clientcmd.ConfigOverrides{})
    if err != nil {
        return nil, err
    }
    return LoadFederationClientset14FromConfig(config)
}
```

And you can then replace lines 100-110 here by `return LoadFederationClientset14FromConfig(config)`

nikhiljindal commented on test/e2e/federation-authn.go:
```diff
@@ -0,0 +1,110 @@
+		return nil, fmt.Errorf("error creating federation clientset: %v", err)
+	}
+	// Set timeout for each client in the set.
+	c.DiscoveryClient.Client.Timeout = framework.SingleCallTimeout
```
We should set the timeout for ExtensionsClient as well.
If you do the refactoring mentioned above, then you will have to fix it at only one place :)

madhusudancs replied:
Added timeouts to `ExtensionsClient`.

madhusudancs commented:
@nikhiljindal Addressed your comments. PTAL.

nikhiljindal commented:
Thanks @madhusudancs 
Added some minor comments (nothing merge blocking).

Feel free to squash commits and add LGTM label.

madhusudancs commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/e2e/federation-authn.go, line 100 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30409#-KOr7rPFg98oVTYKONzv-r1-100:-KPQYrQlgHcVAsphFSoB:b-9hwbod) ([raw file](https://github.com/kubernetes/kubernetes/blob/f1f7f88a12312ae98bdb19908445a15d2c85f48c/test/e2e/federation-authn.go#L100)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Yes. Thats why I suggested refactoring the part after you have the config. Something like LoadFederationClientset14FromConfig(config). So LoadFederationClientset_1_4 becomes:
> 
> ```
> 
> LoadFederationClientset_1_4() {
> 
>     config, err := LoadFederatedConfig(&clientcmd.ConfigOverrides{})
> 
>     if err != nil {
> 
>         return nil, err
> 
>     }
> 
>     return LoadFederationClientset14FromConfig(config)
> 
> }
> 
> ```
> 
> 
> 
> And you can then replace lines 100-110 here by `return LoadFederationClientset14FromConfig(config)`
</details>
I still maintain my position. I would still keep `LoadFederationClientset_1_4()` from this. It's not worth having everyone choose between the two functions all the time. This is quite specialized to keep them separate.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30409)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/e2e/federation-authn.go, line 52 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30409#-KOr7rPFg98oVTYKONzv-r1-52:-KPQaiPezwfxET_4UoXO:b-aqdsrc) ([raw file](https://github.com/kubernetes/kubernetes/blob/f1f7f88a12312ae98bdb19908445a15d2c85f48c/test/e2e/federation-authn.go#L52)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> oh, yes. so we can use any context. makes sense.
> 
> Would be good if you can add that as a comment.
</details>
Added this as a comment in the code.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30409)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@nikhiljindal addressed your comments, also squashed the commits. Adding LGTM label.

Thanks for the review!

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30409#-:-KPQapVs1MuUDw7AlWLH:b-hrgnyt)*
<!-- Sent from Reviewable.io -->
